File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/Conditional.java
//////////////////////////////////////////////////////////////////////////////
// Conditional.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.util.Optional;
import java.util.function.Supplier;

/**
 * <p>
 * Wraps a {@link Boolean} condition and provides a fluent API for branching,
 * combining, and throwing based on its value, as an alternative to
 * imperative {@code if}/{@code else} statements.
 * </p>
 * <br/>
 * <b>Examples</b><br/>
 * <pre>
 * Conditional
 *     .of(user.isActive())
 *     .ifTrue(() -&gt; sendWelcomeEmail(user))
 *     .ifFalse(() -&gt; disableAccount(user));
 * </pre>
 */
public
class Conditional
    implements Supplier<Boolean>,Comparable<Conditional>
{
    public static final Conditional TRUE = Conditional.of(true);
    public static final Conditional FALSE = Conditional.of(false);

    private final Boolean condition;

    public Conditional(Boolean c)
    {
        condition = c;
    }

    @Override
    public int
    hashCode()
    {
        return 73*condition.hashCode();
    }

    @Override
    public boolean
    equals(Object other)
    {
        return
            other instanceof Conditional otherConditional
                ? condition.equals(otherConditional.condition)
                : false;
    }

    @Override
    public Boolean
    get()
    {
        return condition;
    }

    @Override
    public int
    compareTo(Conditional other)
    {
        return condition.compareTo(other.condition);
    }

    public <T> Optional<T>
    ifTrue(Supplier<T> supplier)
    {
        return
            condition
                ? Optional.ofNullable(supplier.get())
                : Optional.empty();
    }

    public <T> Optional<T>
    ifFalse(Supplier<T> supplier)
    {
        return
            !condition
                ? Optional.ofNullable(supplier.get())
                : Optional.empty();
    }

    public void
    ifTrue(Runnable action)
    {
        if (condition)
            action.run();
    }

    public void
    ifFalse(Runnable action)
    {
        if (!condition)
            action.run();
    }

    public <T> T
    ifTrueOrThrow(Supplier<T> supplier,Throwable exception)
        throws Throwable
    {
        if (condition)
            return supplier.get();

        throw exception;
    }

    public <T> T
    ifFalseOrThrow(Supplier<T> supplier,Throwable exception)
        throws Throwable
    {
        if (!condition)
            return supplier.get();

        throw exception;
    }

    public void
    ifTrueOrThrow(Runnable action,Throwable exception)
        throws Throwable
    {
        if (condition)
            action.run();
        else
            throw exception;
    }

    public void
    ifFalseOrThrow(Runnable action,Throwable exception)
        throws Throwable
    {
        if (!condition)
            action.run();
        else
            throw exception;
    }

    public <T> T
    ifTrueOrElse(Supplier<T> trueSupplier,Supplier<T> falseSupplier)
    {
        if (condition)
            return trueSupplier.get();

        return falseSupplier.get();
    }

    public void
    ifTrueOrElse(Runnable trueAction,Runnable falseAction)
    {
        if (condition)
            trueAction.run();
        else
            falseAction.run();
    }

    public Boolean
    orElse(Supplier<Boolean> elseSupplier)
    {
        if (condition)
            return true;

        return elseSupplier.get();
    }


    public Conditional
    and(Conditional other)
    {
        return Conditional.of(condition && other.condition);
    }

    public Conditional
    and(Boolean other)
    {
        return Conditional.of(condition && other);
    }

    public Conditional
    or(Conditional other)
    {
        return Conditional.of(condition || other.condition);
    }

    public Conditional
    or(Boolean other)
    {
        return Conditional.of(condition || other);
    }

    public Conditional
    xor(Conditional other)
    {
        return Conditional.of(condition ^ other.condition);
    }

    public Conditional
    xor(Boolean other)
    {
        return Conditional.of(condition ^ other);
    }

    public Conditional
    not()
    {
        return Conditional.of(!condition);
    }

    public static Conditional
    of(Boolean source)
    {
        return new Conditional(source);
    }

}

//////////////////////////////////////////////////////////////////////////////

File: Components/Strata.Foundation.Core/main/typescript/Utility/Conditional.ts
import {ISupplier} from "./ISupplier";
import { Optional } from "./Optional";
import {ISupplierOrLambda, LambdaSupplier} from "./LambdaSupplier";
import {IRunnableOrLambda, LambdaRunnable} from "./LambdaRunnable";

export
class Conditional
    implements ISupplier<boolean>
{
    public static readonly TRUE: Conditional = Conditional.of(true);
    public static readonly FALSE: Conditional = Conditional.of(false);

    private readonly condition: boolean;

    constructor(condition: boolean)
    {
        this.condition = condition;
    }

    get(): boolean
    {
        return this.condition;
    }

    ifTrue<T>(supplier: ISupplierOrLambda<T>): Optional<T>
    {
        return this.condition
            ? Optional.ofNullable(LambdaSupplier.of(supplier).get())
            : Optional.empty();
    }

    ifFalse<T>(supplier: ISupplierOrLambda<T>): Optional<T>
    {
        return !this.condition
            ? Optional.ofNullable(LambdaSupplier.of(supplier).get())
            : Optional.empty();
    }

    ifTrueNoReturn(action:IRunnableOrLambda)
    {
        if (this.condition)
            LambdaRunnable.of(action).run();
    }

    ifFalseNoReturn(action:IRunnableOrLambda)
    {
        if (!this.condition)
            LambdaRunnable.of(action).run();
    }

    ifTrueOrThrow<T>(supplier: ISupplierOrLambda<T>,error: Error): T
    {
        if (this.condition)
            return LambdaSupplier.of(supplier).get();

        throw error;
    }

    ifFalseOrThrow<T>(supplier: ISupplierOrLambda<T>,error: Error): T
    {
        if (!this.condition)
            return LambdaSupplier.of(supplier).get();

        throw error;
    }

    ifTrueOrThrowNoReturn(action: IRunnableOrLambda,error: Error): void
    {
        if (this.condition)
            LambdaRunnable.of(action).run();
        else
            throw error;
    }

    ifFalseOrThrowNoReturn(action: IRunnableOrLambda,error: Error): void
    {
        if (!this.condition)
            LambdaRunnable.of(action).run();
        else
            throw error;
    }

    ifTrueOrElse<T>(
        trueSupplier: ISupplierOrLambda<T>,
        falseSupplier: ISupplierOrLambda<T>): T
    {
        if (this.condition)
            return LambdaSupplier.of(trueSupplier).get();

        return LambdaSupplier.of(falseSupplier).get();
    }

    ifTrueOrElseNoReturn(
        trueAction: IRunnableOrLambda,
        falseAction: IRunnableOrLambda): void
    {
        if (this.condition)
            LambdaRunnable.of(trueAction).run();
        else
            LambdaRunnable.of(falseAction).run();
    }

    and(other: Conditional|boolean): Conditional
    {
        if (typeof other === 'boolean')
        {
            const input: boolean = other;

            return Conditional.of(this.condition && input);
        }

        return Conditional.of(this.condition && other.condition);
    }

    or(other: Conditional|boolean): Conditional
    {
        if (typeof other === 'boolean')
        {
            const input: boolean = other;

            return Conditional.of(this.condition || input);
        }

        return Conditional.of(this.condition || other.condition);
    }

    xor(other: Conditional|boolean): Conditional
    {
        if (typeof other === 'boolean')
        {
            const input: boolean = other;

            return Conditional.of(this.condition !== input);
        }

        return Conditional.of(this.condition !== other.condition);
    }

    not(): Conditional
    {
        return Conditional.of(!this.condition);
    }

    public static of(source: boolean): Conditional
    {
        return new Conditional(source);
    }

}